File: 238-product-of-array-except-self/product-of-array-except-self.java
class Solution {
    public int[] productExceptSelf(int[] nums) {
        int sprod = 1;
        int nprod = -1;
        int count = 0;
        int n= nums.length;
        for(int i=n-1;i>=0;i--){
            if(nums[i]==0){
                count++;
                nprod=sprod;
            }
            sprod*=nums[i];
        }
        int arr[] = new int[n];
        if(count>1) return arr;
        int pprod = 1;
        for(int i=0;i<n;i++){
            if(nums[i]==0){
                arr[i]=pprod*nprod;
            }
            else{
                arr[i] = sprod/nums[i];
            }
            pprod*=nums[i];
        }
        return arr;
    }
}

File: 238-product-of-array-except-self/product-of-array-except-self.cpp
class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        int n = nums.size();
        vector<int> prefix(n);
        vector<int> suffix(n);
        int prp = 1;
        int srp = 1;
        for(int i=0;i<n;i++){
            prp*=nums[i];
            prefix[i] = prp;
            srp*=nums[n-i-1];
            suffix[n-i-1] = srp;
        }
        vector<int> ans;
        for(int i=0;i<n;i++){
            int pre = i>0 ? prefix[i-1]:1;
            int suf = i<n-1 ? suffix[i+1]:1;
            ans.push_back(pre*suf);
        }
        return ans;
    }
};